scripts/test.sh

```diff
@@ -81,26 +102,118 @@ start_packager() {
 }
 
 xctest() {
-  local dest="$(xcrun simctl list devices | grep -v unavailable | grep -m 1 -o '[0-9A-F\-]\{36\}')"
+  # note: the DEVELOPER_DIR and IOS_SIM_DEVICE sections sould be kept in sync with setup_ios_simulator
+  
+  # - Ensure one version of xcode is chosen by all tools
+  if [[ -z "$DEVELOPER_DIR" ]]; then
+    export DEVELOPER_DIR="$(xcode-select -p)"
+  fi
+  
+  # - Choose a device, if it has not already been chosen
+  local deadline=$((SECONDS+5))
+  while [ -z "$IOS_SIM_DEVICE" ] && [ $SECONDS -lt $deadline ]; do
+    IOS_SIM_DEVICE=`ruby -rjson -e "puts JSON.parse(%x{xcrun simctl list devices --json})['devices'].each{|os,group| group.each{|dev| dev['os'] = os}}.flat_map{|x| x[1]}.select{|x| x['availability'] == '(available)'}.each{|x| x['score'] = (x['name'] == '$ios_sim_default_device_type' ? 1 : 0) + (x['os'] == '$ios_sim_default_ios_version' ? 1 : 0)}.sort_by!{|x| [x['score'], x['name']]}.reverse![0]['udid']"`
+    export IOS_SIM_DEVICE
+  done
+  if [ -z "$IOS_SIM_DEVICE" ]; then
+    echo "*** Failed to determine the iOS Simulator device to use ***"
+    exit 1
+  fi
+  
+  # - Wait until the simulator is fully booted by waiting for it to launch SpringBoard
+  printf "Waiting for springboard to ensure device is ready..."
+  xcrun simctl launch "$IOS_SIM_DEVICE" com.apple.springboard >/dev/null 2>&1 || true
+  echo "  done"
+  
+  # - Run the build and test
   if [ -n "$XCPRETTY" ]; then
-    LOGTEMP=`mktemp build.log.XXXXXX`
-    trap "rm $LOGTEMP" EXIT
-    xcodebuild -scheme "$1" -configuration "$CONFIGURATION" -sdk iphonesimulator -destination id="$dest" build test 2>&1 | tee "$LOGTEMP" | "$XCPRETTY" -c --no-utf --report junit --output build/reports/junit.xml || {
+    log_temp=`mktemp build.log.XXXXXX`
+    if [ -e "$log_temp" ]; then
+      rm "$log_temp"
+    fi
+    xcrun xcodebuild -scheme "$1" -configuration "$CONFIGURATION" -sdk iphonesimulator -destination name="iPhone 5s" build test 2>&1 | tee "$log_temp" | "$XCPRETTY" -c --no-utf --report junit --output build/reports/junit.xml || {
 	    EXITCODE=$?
         printf "*** Xcode Failure (exit code $EXITCODE). The full xcode log follows: ***\n\n"
-        cat "$LOGTEMP"
+        cat "$log_temp"
         printf "\n\n*** End Xcode Failure ***\n"
         exit $EXITCODE
     }
+    rm "$log_temp"
   else
-    xcodebuild -scheme "$1" -configuration "$CONFIGURATION" -sdk iphonesimulator -destination id="$dest" build test || {
+    xcrun xcodebuild -scheme "$1" -configuration "$CONFIGURATION" -sdk iphonesimulator -destination id="$IOS_SIM_DEVICE" build test || {
 	    EXITCODE=$?
         echo "*** Failure (exit code $EXITCODE). ***"
         exit $EXITCODE
-	}
+	  }
   fi
 }
 
+setup_ios_simulator() {
+  # - Ensure one version of xcode is chosen by all tools
+  
+  if [[ -z "$DEVELOPER_DIR" ]]; then
+    export DEVELOPER_DIR="$(xcode-select -p)"
+  fi
+  
+  # - Kill the Simulator to ensure we are running the correct one
```

Comment:
Trying to put that much context in the comments feels a bit odd given that one is in a method called `cleanup`, and the other in `setup_ios_simulator`, with a section comment explaining why we are quitting Simulator.app.

And there are meaningful differences in the way we kill these based on those circumstances (clearing the way before us vs. cleaning up our own mess) that would make trying to combine this awkward.